### PR TITLE
Constrain scalar types with concept

### PR DESCRIPTION
**Pull request by jhale**
This PR constrains T (which is usually used to denote a scalar related to the finite element coefficients) to be a real or complex floating point type. Let me know if this needs to be widened re: multi-precision support.

In terms of naming 'field' might be more mathematically correct than 'scalar'.

I wonder also if we should change all of the Mesh related Ts to Us throughout?

### CODE REVIEW

**Comment by jhale** [comment]
Looking at the diff I've used dolfinx::scalar where instead I should have used std::floating_point so I will make these changes.

File: cpp/dolfinx/fem/Constant.h [hunk 1]
```diff
@@ -10,12 +10,14 @@
 #include <span>
 #include <vector>
 
+#include "dolfinx/common/types.h"
```
Comment (garth-wells) [inline cpp/dolfinx/fem/Constant.h:13]: Remove blank line above.

File: cpp/dolfinx/fem/Constant.h [hunk 2]
```diff
@@ -10,12 +10,14 @@
...
+
 namespace dolfinx::fem
 {
 
 /// Constant value which can be attached to a Form. Constants may be
 /// scalar (rank 0), vector (rank 1), or tensor-valued.
-template <typename T>
+template <dolfinx::scalar T>
```
Comment (garth-wells) [inline cpp/dolfinx/fem/Constant.h:20]: Add `/// @tparam T ...` to docstring.

**Review by garth-wells — COMMENTED** [review]
Looks good. Tiny formatting comment.

**Review by garth-wells — APPROVED** [review]

**Pull request merged**